### ISSUE TITLE
Certain point-wise GPU operations with broadcasting are up to 12x slower going from v0.2.0 to master

### ISSUE THREAD

**Issue by ducksoup** [pytorch/pytorch#2816]
Some point-wise tensor operations are up to 12 times slower on master (https://github.com/pytorch/pytorch/commit/2b9765ad02b666d623566935b385fc3d058ad33d) compared to v0.2.0 when one of the operands is broadcasted.

We are able to consistently reproduce this issue with the following code (note that the same issue appears also for point-wise sum and division):
```python
import torch
import time

N, C, H, W = 64, 256, 64, 64
repetitions = 100

x = torch.randn(N, C, H, W).cuda()
w = torch.randn(C).cuda()

t0 = time.time()
for _ in range(repetitions):
    y = x * w.view(1, C, 1, 1)
    torch.cuda.synchronize()
print (time.time() - t0) / repetitions
```

A typical output is as follows:
```
(pytorch_master) $ python timeit.py 
0.0461528801918

(pytorch_v0.2) $ python timeit.py 
0.00369760036469
```

The slow-down seems to be related to the axis that `x` and `w` have in common, in fact, when using
```python
w = torch.randn(N).cuda()
...
y = x * w.view(N, 1, 1, 1)
```
or
```python
w = torch.randn(W).cuda()
...
y = x * w.view(1, 1, 1, W)
```
there is no noticeable slow-down, while
```python
w = torch.randn(H).cuda()
...
y = x * w.view(1, 1, H, 1)
```
produces the same behavior as the script above.

This has been tested on different systems using CUDA 8.0 and a variety of GPUs, including: TITAN X, TITAN Xp and GTX 1050.

**Comment by gchanan** [comment]
Interesting!  I'll take a look.

**Comment by colesbury** [comment]
It's this commit: 80caca4edbc415abb2f0695fb2565e6b46c410a8

The larger grid size appears to hurt this use case.

**Comment by csarofeen** [comment]
This a CUDA <9 issue. CUDA may be trying to drop the entire THCTensorInfo object into register space (This may happen with -1 indexing). If this is the case it makes sense that smaller blocks would help. @ducksoup can you try... '''import torch
import time

N, C, H, W = 64, 256, 64, 64
repetitions = 100

x = torch.randn(N, C, H, W).cuda()
w = torch.randn(C).cuda().expand(N, C).contiguous()

t0 = time.time()
for _ in range(repetitions):
    y = x * w.view(N, C, 1, 1)
    torch.cuda.synchronize()
print( (time.time() - t0) / repetitions )
'''

**Comment by ducksoup** [comment]
@csarofeen from your code I get the following output (same setup as the one used in my original comment):
```
(pytorch_master) $ python timeit.py 
0.00187160015106
```
I need to set `pytorch_v0.2` back up though, I will edit this with the results when ready.

**Comment by Stonesjtu** [comment]
@csarofeen  Could you add some references why dropping the entire THCTensorInfo into register space slows down this operation?

**Comment by apaszke** [comment]
My guess would be that the # of registers required to run a single block rises dramatically, which means that you can't fill an SM with blocks, because it runs out of registers faster than out of compute resources

**Comment by csarofeen** [comment]
@Stonesjtu Keeping data into register space is not inherently bad. In fact it can often be a very useful optimization which is why the compiler has a tendency to do so. THCTensorInfo, however, requires >200B. https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorInfo.cuh#L46-L49 In THCApply we are trying to put 4, 512 blocks per SM (https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCApply.cuh#L24) Which means that if we try to put THCTensorInfo in register space on every thread we're asking for 409KB of register space. Considering register space / SM is 256KB this clearly can't be stored there. Instead it is stored in thread local memory (gets sent to global memory). http://developer.download.nvidia.com/CUDA/training/register_spilling.pdf

There for every block is storing 200B * 512 threads of memory into global memory before it even gets started. Keeping in mind that each block is responsible for 2 reads and 1 write (assuming THCApply 2) or 12B*512 threads.

The reason that switching from a tensor of (1, C, 1, 1) to (N, C, 1, 1) with first 2 dimensions contiguous helps is that the backend 'collapses' N, C, 1, 1 tensor into a 2D tensor therefore using THCTensor's 2D templated indexing. https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCApply.cuh#L484-L548 which templates https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorInfo.cuh#L228-L248 .

Since Dims in this loop https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/THCTensorInfo.cuh#L236 is a template parameter, the compiler now knows exactly what is needed from the THC tensor structure (not a runtime variable as in -1 case) it will decide to not try and put it in thread local memory.

**Comment by Stonesjtu** [comment]
@csarofeen  URLs are out-dated. Location of pytorch cuda tensors goes to https://github.com/pytorch/pytorch/tree/master/aten/src/THC/. Or you may specify a commit rather than master branch.

**Comment by gchanan** [comment]
closing as an old issue.
